rosbag2_compression/src/rosbag2_compression/zstd_decompressor.cpp

```diff
@@ -0,0 +1,201 @@
+}
+
+void write_output_buffer(
+  const uint8_t * output_buffer,
+  const size_t output_buffer_length,
+  const std::string & uri)
+{
+  const auto file_pointer = open_file(uri.c_str(), "wb");
+  if (file_pointer == nullptr) {
+    std::stringstream errmsg;
+    errmsg << "Failed to open file: \"" << uri << "\" for binary writing!";
+
+    throw std::runtime_error(errmsg.str());
+  }
+
+  const auto nWrite = fwrite(
+    output_buffer, sizeof(uint8_t), output_buffer_length, file_pointer);
+
+  if (nWrite != output_buffer_length) {
+    ROSBAG2_COMPRESSION_LOG_ERROR_STREAM("Bytes written (" <<
+      nWrite << " != output_buffer_length (" << output_buffer_length <<
+      ")!");
+    // An error indicator is set by fwrite, so the following check will throw.
+  }
+
+  if (ferror(file_pointer)) {
+    fclose(file_pointer);
+
+    std::stringstream errmsg;
+    errmsg << "Unable to write decompressed data to file: \"" << uri << "\"!";
+
+    throw std::runtime_error(errmsg.str());
+  }
+
+  fclose(file_pointer);
+}
+
+void throw_on_zstd_error(const size_t compression_result)
+{
+  if (ZSTD_isError(compression_result)) {
+    std::stringstream error;
+    error << "ZSTD decompression error: " << ZSTD_getErrorName(compression_result);
+
+    throw std::runtime_error(error.str());
+  }
+}
+
+void check_frame_content(const size_t frame_content)
+{
+  if (frame_content == ZSTD_CONTENTSIZE_ERROR) {
+    throw std::runtime_error("File not compressed with Zstd.");
+  } else if (frame_content == ZSTD_CONTENTSIZE_UNKNOWN) {
+    throw std::runtime_error("Unable to determine file size");
+  }
+}
+
+void print_decompression_statistics(
+  const std::chrono::high_resolution_clock::time_point start,
+  const std::chrono::high_resolution_clock::time_point end,
+  const size_t decompressed_size,
+  const size_t compressed_size)
+{
+  const auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
+  const auto decompression_ratio =
+    static_cast<double>(decompressed_size) / static_cast<double>(compressed_size);
+
+  ROSBAG2_COMPRESSION_LOG_DEBUG_STREAM(
+    "Decompression statistics:\n" <<
+      "Time: " << duration.count() << " microseconds" <<
+      "Compression Ratio: " << decompression_ratio);
+}
+}  // namespace
+
+namespace rosbag2_compression
+{
+
+std::string ZstdDecompressor::decompress_uri(const std::string & uri)
+{
+  const auto start = std::chrono::high_resolution_clock::now();
+  const auto uri_path = rcpputils::fs::path(uri);
+  const auto decompressed_uri = rcpputils::fs::remove_extension(uri_path).string();
+
+  size_t compressed_buffer_length{0};
+  const auto compressed_buffer = get_input_buffer(uri, compressed_buffer_length);
+
+  const auto decompressed_buffer_length =
+    ZSTD_getFrameContentSize(compressed_buffer.data(), compressed_buffer_length);
+  check_frame_content(decompressed_buffer_length);
+
+  std::vector<uint8_t> decompressed_buffer;
+  decompressed_buffer.reserve(decompressed_buffer_length);
```

Comment:
```suggestion
  decompressed_buffer.resize(decompressed_buffer_length);
```

Reply:
See: https://github.com/ros2/rosbag2/pull/230#discussion_r362152097 for resolution